[Qbus] Initial contribution

QbusKoen opened the pull request:
Thist is te initial contribution for the Qbus binding. This binding has been tested already by multiple users. Especially for openHAB2 an by some people for openHAB(3). The support for this binding was done on the[ Qbus forum](http://www.qbusforum.be/).

This binding does not communicate directly to the home automation controller, but indirectly by a Client/Server application more info can be found here: [https://github.com/QbusKoen/QbusClientServer](https://github.com/QbusKoen/QbusClientServer).

We also made an online user guide, explaining the client/server and the basic principles of openHAB on [Manual openHAB - Qbus](https://manualoh.schockaert.tk/nl/Inleiding)

The JAR file for testing can be found here: [https://github.com/QbusKoen/QbusOH3-JAR](https://github.com/QbusKoen/QbusOH3-JAR)

Signed-off-by: Koen Schockaert ks@qbus.be

fwolter commented:
The link to "Manual openHAB - Qbus" leads to a RPI page.

What is the reason for keeping the "core" of this binding in a separate repository?

QbusKoen commented:
> The link to "Manual openHAB - Qbus" leads to a RPI page.
> 
> What is the reason for keeping the "core" of this binding in a separate repository?

The first page of the manual indeed shows a rpi, as an example of a device capable of runnin openHAB and the Qbus client/server.

The reason for the seperated core is that this core excistetd before i knew about openHAB and was the only way to communicate to the Qbus system. The protocol of the TCP communication with Qbus is a closed protocol for security reasons. Therefore a C# DLL was created to give people with knowledge of using the DLL the possibility to communicate with Qbus.
I created a client/server application from the DLL that allows communication by json data, which makes it easier for people with less programming experience to communicate with Qbus.

So, that's why i decided to use the 'core' as a bridge between Qbus and openHAB.

fwolter commented:
Just to get an idea. Is the client/server application used in other projects? Are there other maintainers than you?

QbusKoen commented:
> Just to get an idea. Is the client/server application used in other projects? Are there other maintainers than you?

The client/server will be used in other projects, but for the moment only in openHAB, and i'm the only maintainer (for the moment - don't know what the future will bring ;-)).

Maybe it is usefull for you to know that i'm an employee of Qbus. I started on the support departement, but with my earlier expieriences in programming and system integration if spend more time in that section. I started experminting to make our system more open. 
So now i'm more the IoT specialist for Qbus rather than a support engineer, although i still offer support on our forum but less for Qbus spefic questions - more for openHAB related questions.

fwolter commented on bundles/org.openhab.binding.qbus/README.md:
```diff
@@ -0,0 +1,110 @@
+
+![Qbus Logo](doc/Logo.JPG)
+
+This binding for Qbus communicates for all controllers of the Qbus home automation system.
+
+More information about Qbus can be found here:
+[Qbus](https://qbus.be)
```
```suggestion
This binding for [Qbus](https://qbus.be) communicates for all controllers of the Qbus home automation system.
```

QbusKoen replied:
Done

fwolter commented on bundles/org.openhab.binding.qbus/README.md:
```diff
@@ -0,0 +1,110 @@
+More information about Qbus can be found here:
+[Qbus](https://qbus.be)
+
+This is the link to the [Qbus forum](https://qbusforum.be). This forum is mainly in dutch and you can find a lot of information about the pre testings af this binding and offers a way to communicate with other users.
```
This could be moved to the end of the readme.

fwolter commented on bundles/org.openhab.binding.qbus/README.md:
```diff
@@ -0,0 +1,110 @@
+
+This is the link to the [Qbus forum](https://qbusforum.be). This forum is mainly in dutch and you can find a lot of information about the pre testings af this binding and offers a way to communicate with other users.
+
+We also host a site which contains a [manual](https://manualoh.schockaert.tk/) where you can find lot's of information.
```
Can you mention that this is in dutch?

Is this related to openHAB or to Qbus in general?

fwolter commented on bundles/org.openhab.binding.qbus/README.md:
```diff
@@ -0,0 +1,110 @@
+
+The controllers can not communicate directly with openHAB, therefore we developed a Client/Server application which you must install prior to enable this binding.
+More information can be found here:
+[Qbus Client/Server](https://github.com/QbusKoen/QbusClientServer)
```
You might want to link to https://github.com/QbusKoen/QbusClientServer-Installer

fwolter commented on bundles/org.openhab.binding.qbus/README.md:
```diff
@@ -0,0 +1,110 @@
+
+The following things are supported by the Qbus Binding:
+
+- Dimmer 1 button, 2 button and clc as _dimmer_
```
```suggestion
- `dimmer`: Dimmer 1 button, 2 button and clc
```

fwolter commented on bundles/org.openhab.binding.qbus/README.md:
```diff
@@ -0,0 +1,110 @@
+The following things are supported by the Qbus Binding:
+
+- Dimmer 1 button, 2 button and clc as _dimmer_
+- Bistabiel, Timer1-3, Interval as _onOff_
```
Is this spelled correctly?

fwolter commented on bundles/org.openhab.binding.qbus/README.md:
```diff
@@ -0,0 +1,110 @@
+
+- Dimmer 1 button, 2 button and clc as _dimmer_
+- Bistabiel, Timer1-3, Interval as _onOff_
+- Thermostats - normal and PID as _thermosats_
```
Any reason this is plural and the others singular?

fwolter commented on bundles/org.openhab.binding.qbus/README.md:
````diff
@@ -0,0 +1,110 @@
+    rollershutter_slats      121   "Roller2"          [ rolId=264 ]
+}
+```
+The Bridge connects to the QbusServer, so if the Client/Server application is installed on the same machine then localhost can be used as address. sn is the serial nr of your controller, port should allways be 8447, except in special installations as it is the communication port of the Server application. refresh is a time in minutes which will check the status of the server and reconnects if con ...
````
Can you replace the configuration parameter text by a table? See here for example https://www.openhab.org/addons/bindings/telegram/

fwolter commented on bundles/org.openhab.binding.qbus/README.md:
```diff
@@ -0,0 +1,110 @@
+
+## Channels
+
+| channel             | type          | description                                             |
```
```suggestion
| Thing Type ID | Channel Name         | description                                             |
```

fwolter replied:
Please add the Item Type of each Channel and if it's read-only or readable/writable.

fwolter commented on bundles/org.openhab.binding.qbus/README.md:
```diff
@@ -0,0 +1,110 @@
+
+| channel             | type          | description                                             |
+|---------------------|---------------|---------------------------------------------------------|
+| onOff               | switch        | This is the channel for Bistable, Timers and Intervals  |
```
```suggestion
| `onOff`               | switch        | This is the channel for Bistable, Timers and Intervals  |
```

fwolter commented on bundles/org.openhab.binding.qbus/README.md:
```diff
@@ -0,0 +1,110 @@
+
+## Full Example
+
+### Things:
```
```suggestion
### Things
```

fwolter commented on bundles/org.openhab.binding.qbus/README.md:
````diff
@@ -0,0 +1,110 @@
+```
+Dimmer              ToonzaalLED          <light>    [ "Lighting" ]      {channel="qbus:dimmer:CTD007841:1:brightness"}
+Switch              Toonzaal230V         <light>                        {channel="qbus:onOff:CTD007841:30:switch"}
+Number:Temperature  ServiceSP"[%.1f °C]" (GroepThermostaten)            {channel="qbus:thermostat:CTD007841:50:setpoint"}
````
```suggestion
Number:Temperature  ServiceSP"[%.1f %unit%]" (GroepThermostaten)            {channel="qbus:thermostat:CTD007841:50:setpoint"}
```

fwolter commented on bundles/org.openhab.binding.qbus/README.md:
```diff
@@ -0,0 +1,110 @@
+# Qbus Binding
```
There are some typos and capitalization mistakes in the readme.

fwolter commented on bundles/org.openhab.binding.qbus/src/main/java/org/openhab/binding/qbus/internal/QbusBindingConstants.java:
```diff
@@ -0,0 +1,98 @@
+            "rollershutter_slats");
+    public static final ThingTypeUID THING_TYPE_THERMOSTAT = new ThingTypeUID(BINDING_ID, "thermostat");
+    // List of all Thing Type UIDs
+    public static final Set<ThingTypeUID> SCENE_THING_TYPES_UIDS = Collections
+            .unmodifiableSet(Stream.of(THING_TYPE_SCENE).collect(Collectors.toSet()));
```
```suggestion
    public static final Set<ThingTypeUID> SCENE_THING_TYPES_UIDS = Set.of(THING_TYPE_SCENE);
```

fwolter commented on bundles/org.openhab.binding.qbus/src/main/java/org/openhab/binding/qbus/internal/QbusBridgeHandler.java:
```diff
@@ -0,0 +1,241 @@
+
+    @Override
+    public void initialize() {
+        logger.debug("QBUS: initializing bridge handler");
```
Can this message be replaced by using the debugger or by increasing the framework's log level? See point 4 https://www.openhab.org/docs/developer/guidelines.html#f-logging

fwolter commented on bundles/org.openhab.binding.qbus/src/main/java/org/openhab/binding/qbus/internal/QbusBridgeHandler.java:
```diff
@@ -0,0 +1,241 @@
+        InetAddress addr = getAddr();
+        int port = getPort();
+
+        logger.debug("Qbus: bridge handler host {}, port {}", addr, port);
```
Can this message be replaced by using the debugger or by increasing the framework's log level? See point 4 https://www.openhab.org/docs/developer/guidelines.html#f-logging

fwolter commented on bundles/org.openhab.binding.qbus/src/main/java/org/openhab/binding/qbus/internal/QbusBridgeHandler.java:
```diff
@@ -0,0 +1,241 @@
+        if (addr != null) {
+            createCommunicationObject(addr, port);
+        } else {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.OFFLINE.COMMUNICATION_ERROR,
+                    "Qbus: cannot resolve bridge IP with hostname " + config.get(CONFIG_HOST_NAME));
```
Isn't this rather a configuration error than a communication error?

"Qbus: " is redundant here as the UI will show the message in context.

fwolter commented on bundles/org.openhab.binding.qbus/src/main/java/org/openhab/binding/qbus/internal/QbusBridgeHandler.java:
```diff
@@ -0,0 +1,241 @@
+
+            updateStatus(ThingStatus.ONLINE);
+
+            Integer refreshInterval = ((Number) config.get(CONFIG_REFRESH)).intValue();
```
Primitive types should be used where possible.

fwolter commented on bundles/org.openhab.binding.qbus/src/main/java/org/openhab/binding/qbus/internal/QbusBridgeHandler.java:
```diff
@@ -0,0 +1,241 @@
+        }
+
+        // This timer will restart the bridge connection periodically
+        logger.debug("Qbus: Checking for Client communication every {} min", refreshInterval);
```
The binding's name will be included in the log messages. Please check all.
```suggestion
        logger.debug("Checking for Client communication every {} min", refreshInterval);
```

fwolter commented on bundles/org.openhab.binding.qbus/src/main/java/org/openhab/binding/qbus/internal/QbusBridgeHandler.java:
```diff
@@ -0,0 +1,241 @@
+                if (!qbusComm.communicationActive()) {
+                    qbusComm = null;
+                    bridgeOffline();
+                    updateStatus(ThingStatus.OFFLINE);
```
Can you set the ThingStatusDetail and a message why this is offline? You could retrieve a message by letting `communicationActive()` throw an exception, instead of returning a boolean.

fwolter commented on bundles/org.openhab.binding.qbus/src/main/java/org/openhab/binding/qbus/internal/QbusBridgeHandler.java:
```diff
@@ -0,0 +1,241 @@
+                    return;
+                }
+
+                // updateStatus(ThingStatus.ONLINE);
```
Can this be removed?

fwolter commented on bundles/org.openhab.binding.qbus/src/main/java/org/openhab/binding/qbus/internal/QbusBridgeHandler.java:
```diff
@@ -0,0 +1,241 @@
+                // updateStatus(ThingStatus.ONLINE);
+                updateStatus(ThingStatus.ONLINE);
+            } else {
+                logger.debug("Qbus: Communication still active");
```
Can this message be replaced by using the debugger or by increasing the framework's log level? See point 4 https://www.openhab.org/docs/developer/guidelines.html#f-logging

Please check all.

fwolter commented on bundles/org.openhab.binding.qbus/src/main/java/org/openhab/binding/qbus/internal/QbusBridgeHandler.java:
```diff
@@ -0,0 +1,241 @@
+        updateStatus(ThingStatus.ONLINE);
+    }
+
+    @Override
+    public boolean isInitialized() {
+        return true;
+    }
```
What's the purpose of this?

fwolter commented on bundles/org.openhab.binding.qbus/src/main/java/org/openhab/binding/qbus/internal/QbusBridgeHandler.java:
```diff
@@ -0,0 +1,241 @@
+        this.refreshTimer = null;
+    }
+
+    @Override
+    public void handleConfigurationUpdate(Map<String, Object> configurationParameters) {
+        Configuration configuration = editConfiguration();
+        for (Entry<String, Object> configurationParmeter : configurationParameters.entrySet()) {
+            configuration.put(configurationParmeter.getKey(), configurationParmeter.getValue());
+        }
+        updateConfiguration(configuration);
+
+        scheduler.submit(() -> {
+
+            updateStatus(ThingStatus.ONLINE);
+
+            Integer refreshInterval = ((Number) configuration.get(CONFIG_REFRESH)).intValue();
+            setupRefreshTimer(refreshInterval);
+        });
+    }
```
Is this really necessary? `initialize()` is called when the configuration has changed.

fwolter commented on bundles/org.openhab.binding.qbus/src/main/java/org/openhab/binding/qbus/internal/QbusBridgeHandler.java:
```diff
@@ -0,0 +1,241 @@
+    public void initialize() {
+        logger.debug("QBUS: initializing bridge handler");
+
+        Configuration config = this.getConfig();
```
You could create a dedicated class with the configuration parameters and read them here by calling `getConfigAs()`. That makes the casting unnecessary.

fwolter reviewed (changes requested):
IMHO, your client/server/mono approach is good to go, as the alternative would be to re-implement the DLLs.

I reviewed 6 of 30 files, that you get an impression what needs to be done.

There are some checkstyle warnings left. You could take a look at `target/code-analysis/report.html`.

EDIT: Please add yourself to the CODEOWNERS file.

QbusKoen commented:
@fwolter Thanks for reviewing my binding!
I've updated my code with your requests and checked and changed the other files.
Pushed the changes to github.
Should I now close this pull request and create a new one?
Please advice. Thx!

fwolter commented:
I will have a look into it soon. This PR will be used for all further review loops concerning this code, so please don't close it.

You need to sign-off all your commits. The necessary commands are listed when you click on "Details" at the DCO check below.

fwolter commented on bundles/org.openhab.binding.qbus/README.md:
```diff
@@ -0,0 +1,114 @@
+
+This binding for [Qbus](https://qbus.be) communicates with all controllers of the Qbus home automation system.
+
+We also host a site which contains a [manual](https://manualoh.schockaert.tk/) where you can find lot's of information to set up openHAB with Qbus client and server (for the moment only in Dutch).
```
```suggestion
We also host a site which contains a [manual](https://manualoh.schockaert.tk/) where you can find lots of information to set up openHAB with Qbus client and server (for the moment only in Dutch).
```

fwolter commented on bundles/org.openhab.binding.qbus/README.md:
```diff
@@ -0,0 +1,114 @@
+
+The controllers can not communicate directly with openHAB, therefore we developed a client/server application which you must install prior to enable this binding.
+More information can be found here:
+[Qbus Client/Server]( https://github.com/QbusKoen/QbusClientServer-Installer)
```
```suggestion
[Qbus Client/Server](https://github.com/QbusKoen/QbusClientServer-Installer)
```

fwolter commented on bundles/org.openhab.binding.qbus/README.md:
````diff
@@ -0,0 +1,114 @@
+
+## Thing Configuration
+
+This is an example of the things configuration file:
+
+```
+Bridge qbus:bridge:CTD001122 [ addr="localhost", sn="001122", port=8447, refresh=10 ] {
+    dimmer                   1     "ToonzaalLED"      [ dimmerId=100 ]
+    onOff                    30    "Toonzaal230V"     [ bistabielId=76 ]
+    thermostat               50    "Service"          [ thermostatId=99 ]
+    scene                    70    "Disco"            [ sceneId=36 ]
+    co2                      100   "Productie"        [ co2Id=26 ]
+    rollershutter            120   "Roller1"          [ rolId=268 ]
+    rollershutter_slats      121   "Roller2"          [ rolId=264 ]
+}
+```
````
This is redundant to the example chapter can therefore be removed here.

fwolter commented on bundles/org.openhab.binding.qbus/README.md:
```diff
@@ -0,0 +1,114 @@
+| `port`    | 8447      | YES      | The communication port of the client/server |
+| `refresh` | 5         | NO       | Refresh time - After x minutes there will be a check if server is still running and if client is still connected. If not - reconnect |
+
+## Channels
+
+| Thing Type ID  | Channel Name         | Read only  | description                                             |
+|---------------------|---------------|---------------------------------------------------------|
+| `onOff`               | switch        | No | This is the channel for Bistable, Timers and Intervals  |
+| `dimmer`              | brightness    | No | This is the channel for Dimmers 1&2 buttons and CLC     |
+| `scene`               | Switch        | No | This is the channel for scenes                          |
+| `co2`                 | co2           | Yes | This is the channel for CO2 sensors                     |
+| `rollershutter`       | rollershutter | No | This is the channel for rollershutters                  | 
+| `rollershutter_slats` | rollershutter | No | This is the channel for rollershutters with slats       |
+| `thermostat`          | setpoint      | No | This is the channel for thermostats setpoint            |
+| `thermostat`          | measured      | Yes | This is the channel for thermostats currenttemp         |
+| `thermostat`          | mode          | No | This is the channel for thermostats mode                |
```
Are these Channels or Things? Either the chapter name, the column names or the description doesn't fit.

fwolter commented on bundles/org.openhab.binding.qbus/README.md:
````diff
@@ -0,0 +1,114 @@
+```
+
+
+| Property   | Default  | Required | Description    |
+|------------|----------|----------|----------------|
+| `addr`    | localhost | YES      | The ip address of the machine where the Qbus Server runs |
+| `sn`      |           | YES      | The serial number of your controller |
+| `port`    | 8447      | YES      | The communication port of the client/server |
+| `refresh` | 5         | NO       | Refresh time - After x minutes there will be a check if server is still running and if client is still connected. If not - reconnect |
````
Can you fix the formatting of the tables? There are dome markdown table beautifiers on the net.

fwolter commented on bundles/org.openhab.binding.qbus/pom.xml:
```diff
@@ -0,0 +1,17 @@
+  <parent>
+    <groupId>org.openhab.addons.bundles</groupId>
+    <artifactId>org.openhab.addons.reactor.bundles</artifactId>
+    <version>3.0.0-SNAPSHOT</version>
```
To make the binding compile, you need to rebase your branch and set the version to the current snapshot: `3.1.0-SNAPSHOT`.

Here are the commands for rebasing your branch:

If not already done, add the upstream openHAB addon repo as a remote to your local repo and fetch it:
```
git remote add upstream https://github.com/openhab/openhab-addons.git
git fetch upstream
```
Then, you can rebase your PR's branch onto `main`:
```
git rebase upstream/main
```
Finally force-push the rebased branch to this PR's branch:
```
git push origin [your branch name of this PR] --force-with-lease
```

fwolter commented on bundles/org.openhab.binding.qbus/src/main/java/org/openhab/binding/qbus/internal/QbusBridgeHandler.java:
```diff
@@ -0,0 +1,295 @@
+     */
+    @Override
+    public void initialize() {
+        logger.info("Initializing bridge handler");
```
Can this message be replaced by using the debugger or by increasing the framework's log level? See point 4 https://www.openhab.org/docs/developer/guidelines.html#f-logging Please check all.

Also logging to info should be used sparsely e.g. a newly started component or a user file that has been loaded.

fwolter commented on bundles/org.openhab.binding.qbus/src/main/java/org/openhab/binding/qbus/internal/QbusBridgeHandler.java:
```diff
@@ -0,0 +1,295 @@
+        }
+
+        // This timer will check connection with server and client periodically
+        logger.info("Check communication with Server and Client every {} min", refreshInterval);
```
See above. Logging to info. Please check all.

fwolter commented on bundles/org.openhab.binding.qbus/src/main/java/org/openhab/binding/qbus/internal/QbusBridgeHandler.java:
```diff
@@ -0,0 +1,295 @@
+                        bridgeOffline("No connection with Qbus Server");
+                        return;
+                    }
+
+                } else {
+
```
Please remove the empty lines.

fwolter commented on bundles/org.openhab.binding.qbus/src/main/java/org/openhab/binding/qbus/internal/QbusBridgeHandler.java:
```diff
@@ -0,0 +1,295 @@
+        // This timer will check connection with server and client periodically
+        logger.info("Check communication with Server and Client every {} min", refreshInterval);
+        refreshTimer = scheduler.scheduleWithFixedDelay(() -> {
+            logger.info("Checking connection with Qbus Server & Client.");
```
Please remove these kind of log messages. I strongly encourage you to use the debugger to check where your program runs. This saves a lot of headache.

fwolter commented on bundles/org.openhab.binding.qbus/src/main/java/org/openhab/binding/qbus/internal/QbusBridgeHandler.java:
```diff
@@ -0,0 +1,295 @@
+    /**
+     * Update the configuration parameters
+     */
+    @Override
+    public void handleConfigurationUpdate(Map<String, Object> configurationParameters) {
+        Configuration configuration = editConfiguration();
+        for (Entry<String, Object> configurationParmeter : configurationParameters.entrySet()) {
+            configuration.put(configurationParmeter.getKey(), configurationParmeter.getValue());
+        }
+        updateConfiguration(configuration);
+        updateStatus(ThingStatus.ONLINE);
+    }
```
What's the purpose of this method?

fwolter commented on bundles/org.openhab.binding.qbus/src/main/java/org/openhab/binding/qbus/internal/QbusBridgeHandler.java:
```diff
@@ -0,0 +1,295 @@
+    /**
+     * Sets the configuration prameters
+     */
+    protected synchronized void setConfig() {
```
Is there any reason why this is synchronized?

The name `setConfig()` seems a bit misleading. Would `readConfig()` be better?

fwolter commented on bundles/org.openhab.binding.qbus/src/main/java/org/openhab/binding/qbus/internal/QbusBridgeHandler.java:
```diff
@@ -0,0 +1,295 @@
+     * @return the addr
+     */
+
+    @SuppressWarnings("null")
+    public @Nullable InetAddress getAddr() {
+        InetAddress addr = null;
+        try {
+            addr = InetAddress.getByName(config.addr);
+        } catch (UnknownHostException e) {
+            logger.debug("Cannot resolve hostname {} to IP adress", config.addr);
+        }
+        return addr;
+    }
```
You could throw the `UnknownHostException` and handle it in `initialize()`. That would make the null suppression unnecessary. In fact that would make the whole method unnecessary. It's better to throw an exception in case of an error instead of returning null. This way you have the error message in the calling method and can integrate it in the message the user gets presented.

fwolter commented on bundles/org.openhab.binding.qbus/src/main/java/org/openhab/binding/qbus/internal/QbusBridgeHandler.java:
```diff
@@ -0,0 +1,295 @@
+     *
+     * @return the port
+     */
+    @SuppressWarnings("null")
+    public int getPort() {
+        return config.port;
+    }
```
The suppression annotation should be used only if the compiler warning is wrong.
```suggestion
    public @Nullable int getPort() {
        return config.port;
    }
```

QbusKoen replied:
If i follow your sollution, then i get:
The nullness annotation @Nullable is not applicable for the primitive type int

fwolter replied:
`config` can be null here. You need to store `config` to a local variable and do a null check on that.

fwolter commented on bundles/org.openhab.binding.qbus/src/main/java/org/openhab/binding/qbus/internal/QbusBridgeHandler.java:
```diff
@@ -0,0 +1,295 @@
+
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+        // TODO Auto-generated method stub
```
Can this comment be removed?

fwolter reviewed (changes requested):
Here is the review feedback for the readme and the first few code files.

There are some checkstyle warnings left. You could take a look at `target/code-analysis/report.html`.

QbusKoen commented:
@fwolter Thanks for reviewing my binding again!
I've updated my code with your requests.
Just one thing i can't figure out. Left a comment above.

fwolter commented:
Can you please click on "Resolve conversation" on those you fixed?

The DCO check failed. You need to sign-off your commits. See https://www.openhab.org/docs/developer/contributing.html#sign-your-work The necessary commands are listed when you click on "Details" at the DCO check below.

QbusKoen commented:
@fwolter 
It took me some time to make the requested changes to my code, but i completed them now.
Can you check again if you find the time?

Many Thanks!

fwolter commented on bundles/org.openhab.binding.qbus/src/main/java/org/openhab/binding/qbus/internal/handler/QbusBistabielHandler.java:
```diff
@@ -0,0 +1,214 @@
+    private void handleSwitchCommand(QbusBistabiel QBistabiel, ChannelUID channelUID, Command command) {
+        if (command instanceof OnOffType) {
+            OnOffType s = (OnOffType) command;
+            @Nullable
+            String snr = getSN();
```
The compiler is more intelligent on local variables. Therefore, the Nullable could be removed. Please check all.
```suggestion
            String snr = getSN();
```

fwolter commented on bundles/org.openhab.binding.qbus/src/main/java/org/openhab/binding/qbus/internal/handler/QbusCO2Handler.java:
```diff
@@ -0,0 +1,185 @@
+@NonNullByDefault
+public class QbusCO2Handler extends QbusGlobalHandler {
+
+    public QbusCO2Handler(Thing thing) {
+        super(thing);
+    }
+
+    protected @NonNullByDefault({}) QbusThingsConfig config;
```
See above. Constructor below fields and null annotation. Please check all.

fwolter commented on bundles/org.openhab.binding.qbus/src/main/java/org/openhab/binding/qbus/internal/handler/QbusCO2Handler.java:
```diff
@@ -0,0 +1,185 @@
+
+    protected @NonNullByDefault({}) QbusThingsConfig config;
+
+    int co2Id = 0;
+
+    @Nullable
+    String sn;
```
Can these be private?
```suggestion
    private int co2Id = 0;

    private @Nullable String sn;
```

fwolter commented on bundles/org.openhab.binding.qbus/src/main/java/org/openhab/binding/qbus/internal/handler/QbusDimmerHandler.java:
```diff
@@ -0,0 +1,307 @@
+                        switch (channelUID.getId()) {
+                            case CHANNEL_SWITCH:
+                                handleSwitchCommand(QDimmer, command);
+                                updateStatus(ThingStatus.ONLINE);
```
Do you set the Thing to online by intention?

fwolter commented on bundles/org.openhab.binding.qbus/src/main/java/org/openhab/binding/qbus/internal/handler/QbusDimmerHandler.java:
```diff
@@ -0,0 +1,307 @@
+                                break;
+
+                            default:
+                                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR,
+                                        "Channel unknown " + channelUID.getId());
```
It seems a bit unreasonable to set the Thing offline, when the user sent an invalid command to it. Same for the others.

fwolter commented on bundles/org.openhab.binding.qbus/src/main/java/org/openhab/binding/qbus/internal/handler/QbusRolHandler.java:
```diff
@@ -0,0 +1,357 @@
+    private void handleScreenposCommand(QbusRol QRol, Command command) {
+        @Nullable
+        String snr = getSN();
+        if (command instanceof org.openhab.core.library.types.UpDownType) {
```
You could add the type to the imports. Same for below.
```suggestion
        if (command instanceof UpDownType) {
```

fwolter commented on bundles/org.openhab.binding.qbus/src/main/java/org/openhab/binding/qbus/internal/handler/QbusRolHandler.java:
```diff
@@ -0,0 +1,357 @@
+                }
+            }
+        } else if (command instanceof IncreaseDecreaseType)
+
```
Please remove this empty line.

fwolter commented on bundles/org.openhab.binding.qbus/src/main/java/org/openhab/binding/qbus/internal/protocol/QbusBistabiel.java:
```diff
@@ -0,0 +1,101 @@
+@NonNullByDefault
+public final class QbusBistabiel {
+
+    @Nullable
+    private QbusCommunication QComm;
```
Please check all.
```suggestion
    private @Nullable QbusCommunication QComm;
```

fwolter commented on bundles/org.openhab.binding.qbus/src/main/java/org/openhab/binding/qbus/internal/protocol/QbusCommunication.java:
```diff
@@ -0,0 +1,926 @@
+    @Nullable
+    private String CTD;
+
+    private Boolean CTDConnected = false;
```
Primitive types should be used where possible.
```suggestion
    private boolean CTDConnected = false;
```

fwolter commented on bundles/org.openhab.binding.qbus/src/main/java/org/openhab/binding/qbus/internal/protocol/QbusCommunication.java:
```diff
@@ -0,0 +1,926 @@
+        try {
+
+            for (int i = 1; qEventsRunning && (i <= 5); i++) {
+                Thread.sleep(1000);
```
What's the purpose of this sleep?

fwolter commented on bundles/org.openhab.binding.qbus/src/main/java/org/openhab/binding/qbus/internal/protocol/QbusCommunication.java:
```diff
@@ -0,0 +1,926 @@
+                Thread.sleep(1000);
+            }
+            if (qEventsRunning) {
+                logger.error("Starting from thread {}, but previous connection still active after 5000ms",
+                        Thread.currentThread().getId());
```
You could add this message to the IOException. Then, you only need one logging statement for all exceptions in the catch block.

fwolter commented on bundles/org.openhab.binding.qbus/src/main/java/org/openhab/binding/qbus/internal/protocol/QbusCommunication.java:
```diff
@@ -0,0 +1,926 @@
+            }
+
+            if (handler == null) {
+                throw new IOException();
```
You could add a message to give the user a clue what's wrong.

fwolter commented on bundles/org.openhab.binding.qbus/src/main/java/org/openhab/binding/qbus/internal/protocol/QbusCommunication.java:
```diff
@@ -0,0 +1,926 @@
+                qSocket = socket;
+                qOut = new PrintWriter(socket.getOutputStream(), true);
+                qIn = new BufferedReader(new InputStreamReader(socket.getInputStream()));
+                logger.info("Connected via local port {} from thread {}", socket.getLocalPort(),
```
Logging to info should be used sparsely e.g. a newly started component or a user file that has been loaded. This could be debug.

fwolter commented on bundles/org.openhab.binding.qbus/src/main/java/org/openhab/binding/qbus/internal/protocol/QbusCommunication.java:
```diff
@@ -0,0 +1,926 @@
+            // If Qbus Client is connected then initialize, else put Bridge offline
+            if (CTDConnected == true) {
+                initialize();
+                (new Thread(qEvents)).start();
```
Threads created by bindings should be named and set to daemon mode. You can achieve this by using `NamedThreadFactory`. Here are some examples: https://github.com/openhab/openhab-addons/search?q=namedthreadfactory

fwolter commented on bundles/org.openhab.binding.qbus/src/main/java/org/openhab/binding/qbus/internal/protocol/QbusCommunication.java:
```diff
@@ -0,0 +1,926 @@
+            } else {
+                handler.bridgeOffline("No communication with Qbus client");
+            }
+        } catch (IOException | InterruptedException e) {
```
The InterruptedException is thrown when OH is shut-down. That means the current operation should be finished as fast as possible and no further operations should be done. You can achieve this by not catching the exception here but throw the InterruptedException and catch it at the highest level you can. As this is not an error, it shouldn't be logged, but only handled.

fwolter commented on bundles/org.openhab.binding.qbus/src/main/java/org/openhab/binding/qbus/internal/protocol/QbusCommunication.java:
```diff
@@ -0,0 +1,926 @@
+     * Qbus outputs. It is started after initialization of the communication.
+     *
+     */
+    private Runnable qEvents = () -> {
```
You could make this a private method. Then, you can use it when creating the thread by `this::myMethod`.
```suggestion
    private void myMethod() {
```

fwolter commented on bundles/org.openhab.binding.qbus/src/main/java/org/openhab/binding/qbus/internal/protocol/QbusCommunication.java:
```diff
@@ -0,0 +1,926 @@
+    private Runnable qEvents = () -> {
+        String qMessage;
+
+        logger.info("Listening for events on thread {}", Thread.currentThread().getId());
```
Logging to info should be used sparsely e.g. a newly started component or a user file that has been loaded. This could be debug.

Please check all.

QbusKoen replied:
Is it ok to use logge.trace here?

fwolter replied:
Yes.

QbusKoen replied:
:-) THX!

fwolter replied:
I still see `info`.

fwolter replied:
Can you work on this?

fwolter commented on bundles/org.openhab.binding.qbus/src/main/java/org/openhab/binding/qbus/internal/protocol/QbusCommunication.java:
```diff
@@ -0,0 +1,926 @@
+
+        try {
+            if (reader == null) {
+                throw new IOException();
```
See above. Add a message. Please check all.

fwolter commented on bundles/org.openhab.binding.qbus/src/main/java/org/openhab/binding/qbus/internal/protocol/QbusCommunication.java:
```diff
@@ -0,0 +1,926 @@
+            if (reader == null) {
+                throw new IOException();
+            }
+            while (!listenerStopped & ((qMessage = reader.readLine()) != null)) {
```
You could use the built-in interrupted flag. The thread is interrupted when you call `thread.interrupt()`.
```suggestion
            while (!Thread.currentThread().isInterrupted() & ((qMessage = reader.readLine()) != null)) {
```

fwolter commented on bundles/org.openhab.binding.qbus/src/main/java/org/openhab/binding/qbus/internal/protocol/QbusCommunication.java:
```diff
@@ -0,0 +1,926 @@
+
+        qEventsRunning = false;
+
+        logger.warn("Event listener thread stopped on thread {}", Thread.currentThread().getId());
```
Stopping a thread is a normal operation when OH is shut-down. You could make this debug or trace.

fwolter commented on bundles/org.openhab.binding.qbus/src/main/java/org/openhab/binding/qbus/internal/protocol/QbusCommunication.java:
```diff
@@ -0,0 +1,926 @@
+            // Delay after sending data to improve scene execution
+            try {
+                TimeUnit.MILLISECONDS.sleep(250);
+            } catch (InterruptedException e) {
```
See above. Throw InterruptedException.

fwolter replied:
Can you work on this?

fwolter commented on bundles/org.openhab.binding.qbus/src/main/java/org/openhab/binding/qbus/internal/protocol/QbusCommunication.java:
```diff
@@ -0,0 +1,926 @@
+                return;
+            }
+        } else {
+            logger.error("Initialization error");
```
See above. Logging to error. Please check all.

fwolter commented on bundles/org.openhab.binding.qbus/src/main/java/org/openhab/binding/qbus/internal/protocol/QbusCommunication.java:
```diff
@@ -0,0 +1,926 @@
+     *
+     * @param outputs
+     */
+    private void cmdlistscenes(@Nullable List<Map<String, String>> outputs) {
```
```suggestion
    private void cmdListScenes(@Nullable List<Map<String, String>> outputs) {
```

fwolter commented on bundles/org.openhab.binding.qbus/src/main/java/org/openhab/binding/qbus/internal/protocol/QbusMessageDeserializer.java:
```diff
@@ -0,0 +1,111 @@
+
+            return message;
+
+        } catch (IllegalStateException | ClassCastException e) {
```
It's bad practice to catch `ClassCastException`s. You could check the type beforehand with `instanceof`.

fwolter commented on bundles/org.openhab.binding.qbus/src/main/java/org/openhab/binding/qbus/internal/protocol/QbusThermostat.java:
```diff
@@ -0,0 +1,165 @@
+    private QbusCommunication qComm;
+
+    private String id;
+    private Double measured = 0.0;
+    private Double setpoint = 0.0;
+    private Integer mode = 0;
```
Primitive types should be used where possible.

fwolter commented on bundles/org.openhab.binding.qbus/src/main/resources/OH-INF/binding/binding.xml:
```diff
@@ -0,0 +1,10 @@
+	xsi:schemaLocation="https://openhab.org/schemas/binding/v1.0.0 https://openhab.org/schemas/binding-1.0.0.xsd">
+
+	<name>Qbus Binding</name>
+	<description>This is the binding for Qbus.</description>
```
Could you be a bit more verbose, so that somebody who hasn't heard about it gets a clue what this is about?

fwolter commented on bundles/org.openhab.binding.qbus/src/main/resources/OH-INF/binding/binding.xml:
```diff
@@ -0,0 +1,10 @@
+
+	<name>Qbus Binding</name>
+	<description>This is the binding for Qbus.</description>
+	<author>Koen Schockaert</author>
```
The author tag is deprecated and should therefore be removed. See openhab/openhab-core#1844.

fwolter commented on bundles/org.openhab.binding.qbus/src/main/resources/OH-INF/i18n/qbus_xx_XX.properties:
```diff
@@ -0,0 +1,17 @@
+# FIXME: please substitute the xx_XX with a proper locale, ie. de_DE
```
Can this be removed or filled in?

fwolter commented on bundles/org.openhab.binding.qbus/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,266 @@
+				<context>network-address</context>
+			</parameter>
+			<parameter name="sn" type="text" required="true">
+				<label>Serial nr</label>
```
Words in labels should be capitalized (except prepositions and so on). See https://www.openhab.org/docs/developer/bindings/thing-xml.html#formatting-labels-and-descriptions

fwolter replied:
Can you work on this? Please check all.

fwolter commented on bundles/org.openhab.binding.qbus/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,266 @@
+				<label>IP or Host Name</label>
+				<description>IP Address of Qbus server, usually 'localhost'</description>
+				<default>localhost</default>
+				<advanced>false</advanced>
```
`false` is the default value. So, this tag could be removed. Please check all.

fwolter commented on bundles/org.openhab.binding.qbus/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,266 @@
+		<label>Switch</label>
+		<description>Bistabiel-Mono-Timer-Interval output</description>
+		<channels>
+			<channel id="switch" typeId="switch"/>
```
There are some predefined channel types. Then, you don't need to define your own. See https://www.openhab.org/docs/developer/bindings/thing-xml.html#system-state-channel-types
```suggestion
			<channel id="switch" typeId="system.power"/>
```

fwolter commented on bundles/org.openhab.binding.qbus/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,266 @@
+		<label>Dimmer</label>
+		<description>Qbus dimmer output</description>
+		<channels>
+			<channel id="brightness" typeId="brightness"/>
```
See above. There's also a predefined channel for brightness.

fwolter commented on bundles/org.openhab.binding.qbus/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,266 @@
+		<category>Number</category>
+		<state>
+			<options>
+				<option value="0">Manueel</option>
+				<option value="1">Vorst</option>
+				<option value="2">Economisch</option>
+				<option value="3">Comfort</option>
+				<option value="4">Nacht</option>
```
These should be in English. You could add a translation in the file I asked you to remove or fill.

fwolter reviewed (changes requested):
There are some checkstyle warnings left. You could take a look at `target/code-analysis/report.html`.

fwolter commented:
Please note, that your sign-off message must contain your real name.

QbusKoen commented:
Hi @fwolter, i've updated my code with your proposals and pushed changes to git.
If you would be so kind to review the code once more...
Thx!

fwolter commented on bundles/org.openhab.binding.qbus/src/main/java/org/openhab/binding/qbus/internal/QbusBridgeHandler.java:
```diff
@@ -0,0 +1,294 @@
+
+            setBridgeCallBack();
+
+            if (qbusComm != null) {
+                try {
+                    qbusComm.startCommunication();
+                } catch (InterruptedException | IOException e) {
+                    logger.warn("Error on restaring communication.");
+                }
+            }
+
+            if (qbusComm != null) {
+                if (!qbusComm.communicationActive()) {
+                    qbusComm = null;
+                    bridgeOffline("No communication with Qbus Server");
+                    return;
+                }
+            }
+
+            if (qbusComm != null) {
+                if (!qbusComm.clientConnected()) {
+
+                    qbusComm = null;
+                    bridgeOffline("No communication with Qbus Client");
+                    return;
+                }
+            }
```
It seems these can all be handled in the same `if`.

fwolter commented on bundles/org.openhab.binding.qbus/src/main/java/org/openhab/binding/qbus/internal/QbusBridgeHandler.java:
```diff
@@ -0,0 +1,294 @@
+                try {
+                    qbusComm.startCommunication();
+                } catch (InterruptedException | IOException e) {
+                    logger.warn("Error on restaring communication.");
```
It's good practice to append the exception's message to the log message. Should the Thing set to OFFLINE here?
```suggestion
                    logger.warn("Error on restaring communication: {}", e.getMessage());
```

fwolter commented on bundles/org.openhab.binding.qbus/src/main/java/org/openhab/binding/qbus/internal/QbusHandlerFactory.java:
```diff
@@ -0,0 +1,73 @@
+ */
+
+@Component(service = ThingHandlerFactory.class, configurationPid = "binding.qbus")
+
```
Please remove this empty line.

fwolter commented on bundles/org.openhab.binding.qbus/src/main/java/org/openhab/binding/qbus/internal/handler/QbusDimmerHandler.java:
```diff
@@ -0,0 +1,295 @@
+            return;
+        }
+        this.sn = QBridgeHandler.getSn();
+        ;
```
Please remove this colon.

fwolter commented on bundles/org.openhab.binding.qbus/src/main/java/org/openhab/binding/qbus/internal/handler/QbusDimmerHandler.java:
```diff
@@ -0,0 +1,295 @@
+     * Executes the switch command
+     */
+    private void handleSwitchCommand(QbusDimmer QDimmer, Command command) {
+        if (command instanceof OnOffType) {
+            OnOffType s = (OnOffType) command;
+            String snr = getSN();
+
+            if (s == OnOffType.OFF) {
```
You could directly check for the type. Same for below.
```suggestion
            String snr = getSN();

            if (command == OnOffType.OFF) {
```

fwolter commented on bundles/org.openhab.binding.qbus/src/main/java/org/openhab/binding/qbus/internal/handler/QbusRolHandler.java:
```diff
@@ -0,0 +1,346 @@
+            return;
+        }
+        this.sn = QBridgeHandler.getSn();
+        ;
```
Please remove.

fwolter commented on bundles/org.openhab.binding.qbus/src/main/java/org/openhab/binding/qbus/internal/handler/QbusSceneHandler.java:
```diff
@@ -0,0 +1,214 @@
+            return;
+        }
+        this.sn = QBridgeHandler.getSn();
+        ;
```
Please check all.

fwolter commented on bundles/org.openhab.binding.qbus/src/main/java/org/openhab/binding/qbus/internal/handler/QbusThermostatHandler.java:
```diff
@@ -0,0 +1,244 @@
+        String snr = getSN();
+        if (command instanceof QuantityType<?>) {
+            QuantityType<?> s = (QuantityType<?>) command;
+            double sp = s.doubleValue();
```
You need to convert the temperature value you expect. If the user has configured Fahrenheit, `s.doubleValue()` would return the Fahrenheit value. You can call `c.toUnit()` for the conversion.

fwolter commented on bundles/org.openhab.binding.qbus/src/main/java/org/openhab/binding/qbus/internal/protocol/QbusCommunication.java:
```diff
@@ -0,0 +1,913 @@
+        // offline
+        if (CTDConnected == true) {
+            initialize();
+            Thread thread = new Thread(this::qEvents);
+            thread.setName("OH-binding");
+            thread.setDaemon(true);
+            thread.start();
```
The Thread's name should include the Thing UID. You could use `NamedThreadFactory` to retrieve a correctly configured Thread. Here are some examples: https://github.com/openhab/openhab-addons/search?q=NamedThreadFactory

fwolter commented on bundles/org.openhab.binding.qbus/src/main/java/org/openhab/binding/qbus/internal/protocol/QbusCommunication.java:
```diff
@@ -0,0 +1,913 @@
+     * @throws IOException
+     *
+     */
+    public synchronized void stopCommunication() {
```
The readers and writers need to be closed here.

fwolter commented on bundles/org.openhab.binding.qbus/src/main/java/org/openhab/binding/qbus/internal/protocol/QbusCommunication.java:
```diff
@@ -0,0 +1,913 @@
+            if (reader == null) {
+                throw new IOException("Bufferreader for incomming messages not initialized.");
+            }
+            while (!Thread.currentThread().isInterrupted() & ((qMessage = reader.readLine()) != null)) {
```
```suggestion
            while (!Thread.currentThread().isInterrupted() && ((qMessage = reader.readLine()) != null)) {
```

fwolter commented on bundles/org.openhab.binding.qbus/src/main/java/org/openhab/binding/qbus/internal/protocol/QbusCommunication.java:
```diff
@@ -0,0 +1,913 @@
+        } catch (IOException e) {
+            if (!listenerStopped) {
+                qEventsRunning = false;
+                logger.error("Qbus: IO error in listener on thread {}", Thread.currentThread().getId());
```
Bindings should only log to error if something severe happened, like the detection of a bug in your code. This could be debug. See this link for a description of the log levels: https://www.openhab.org/docs/developer/guidelines.html#f-logging

fwolter commented on bundles/org.openhab.binding.qbus/src/main/java/org/openhab/binding/qbus/internal/protocol/QbusCommunication.java:
```diff
@@ -0,0 +1,913 @@
+        if (writer != null) {
+            writer.println(json);
+            // Delay after sending data to improve scene execution
+            try {
+                TimeUnit.MILLISECONDS.sleep(250);
+            } catch (InterruptedException e) {
+                throw new InterruptedException(e.toString());
+            }
```
You can simply remove the try/catch clause to throw the exception.

fwolter commented on bundles/org.openhab.binding.qbus/src/main/java/org/openhab/binding/qbus/internal/protocol/QbusCommunication.java:
```diff
@@ -0,0 +1,913 @@
+
+        try {
+            startCommunication();
+        } catch (InterruptedException | IOException e) {
```
The `InterruptedException` is thrown when OH is shut-down. That means the current operation should be finished as fast as possible and no further operations should be done. You can achieve this by not catching the exception here but throw the InterruptedException and catch it at the highest level you can. As this is not an error, it shouldn't be logged, but only handled.

fwolter replied:
You need to remove the `InterruptedException` from the catch clause to let it be thrown.

fwolter replied:
Can you work on this?

fwolter commented on bundles/org.openhab.binding.qbus/src/main/java/org/openhab/binding/qbus/internal/protocol/QbusCommunication.java:
```diff
@@ -0,0 +1,913 @@
+     * Called by other methods to Qbus server and read response
+     */
+    private void sendAndReadMessage(String command) throws IOException, InterruptedException {
+        @Nullable
```
The compiler is more intelligent on local variables. Therefore, the Nullable could be removed.

fwolter commented on bundles/org.openhab.binding.qbus/src/main/java/org/openhab/binding/qbus/internal/protocol/QbusCommunication.java:
```diff
@@ -0,0 +1,913 @@
+     *
+     */
+    private void Connect() throws InterruptedException, IOException {
+        @Nullable
```
See above. Can be removed. Please check all.

fwolter commented on bundles/org.openhab.binding.qbus/src/main/java/org/openhab/binding/qbus/internal/protocol/QbusCommunication.java:
```diff
@@ -0,0 +1,913 @@
+     * @throws IOException
+     *
+     */
+    private void Connect() throws InterruptedException, IOException {
```
```suggestion
    private void connect() throws InterruptedException, IOException {
```

fwolter commented on bundles/org.openhab.binding.qbus/src/main/java/org/openhab/binding/qbus/internal/protocol/QbusCommunication.java:
```diff
@@ -0,0 +1,913 @@
+                        qBistabiel.setState(state);
+                    }
+                } else {
+                    logger.error("Error in json for BistabBistabiel/Timers/Monos/Intervals");
```
See above. Logging to error. Please check all.

fwolter commented on bundles/org.openhab.binding.qbus/src/main/java/org/openhab/binding/qbus/internal/protocol/QbusCommunication.java:
```diff
@@ -0,0 +1,913 @@
+     *
+     * @param outputs
+     */
+    private void cmdListscenes(@Nullable List<Map<String, String>> outputs) {
```
```suggestion
    private void cmdListScenes(@Nullable List<Map<String, String>> outputs) {
```

fwolter commented on bundles/org.openhab.binding.qbus/src/main/java/org/openhab/binding/qbus/internal/protocol/QbusCommunication.java:
```diff
@@ -0,0 +1,913 @@
+     *
+     * @param data
+     */
+    private void cmdlistrol(@Nullable List<Map<String, String>> outputs) {
```
Please check all.
```suggestion
    private void cmdListRol(@Nullable List<Map<String, String>> outputs) {
```

fwolter commented on bundles/org.openhab.binding.qbus/src/main/java/org/openhab/binding/qbus/internal/protocol/QbusCommunication.java:
```diff
@@ -0,0 +1,913 @@
+                        logger.warn("Bistabiel in controller not known {}", id);
+                        return;
+                    }
+                    logger.info("Event execute bistabiel {} with state {}", id, value1);
```
Logging to info should be used sparsely e.g. a newly started component or a user file that has been loaded. This could be debug.

fwolter commented on bundles/org.openhab.binding.qbus/src/main/java/org/openhab/binding/qbus/internal/protocol/QbusCommunication.java:
```diff
@@ -0,0 +1,913 @@
+                    return;
+                }
+                if (Qdimmer != null) {
+                    logger.info("Event execute dimmer {} with state {}", id, value);
```
See above. Logging to info. Please check all.

fwolter commented on bundles/org.openhab.binding.qbus/src/main/java/org/openhab/binding/qbus/internal/protocol/QbusDimmer.java:
```diff
@@ -0,0 +1,120 @@
+        if (comm != null) {
+            try {
+                comm.sendMessage(QCmd);
+            } catch (InterruptedException e) {
+                logger.warn("Could not send command for dimmer {}", this.id);
```
See above. Handle InterruptedException correctly. Please check all.

fwolter commented on bundles/org.openhab.binding.qbus/src/main/java/org/openhab/binding/qbus/internal/protocol/QbusMessageDeserializer.java:
```diff
@@ -0,0 +1,110 @@
+            return message;
+
+        } catch (IllegalStateException e) {
+            throw new JsonParseException("Unexpected Json type");
```
It's good practice to include the cause in the exception. What throws a `IllegalStateException` anyway?
```suggestion
            throw new JsonParseException("Unexpected Json type", e);
```

fwolter commented on bundles/org.openhab.binding.qbus/src/main/resources/OH-INF/i18n/qbus_nl.properties:
```diff
@@ -0,0 +1,99 @@
+# FIXME: please substitute the xx_XX with a proper locale, ie. de_DE
+# FIXME: please do not add the file to the repo if you add or change no content
```
Can these be removed?

fwolter commented on bundles/org.openhab.binding.qbus/src/main/resources/OH-INF/i18n/qbus_nl.properties:
```diff
@@ -0,0 +1,99 @@
+
+channel-type.qbus.slats.label = Lamellen Bediening
+channel-type.qbus.slats.description = Lamellen bediening van de uitgangen
+
+
+
+
```
Please remove some of these empty lines.

fwolter commented on bundles/org.openhab.binding.qbus/src/main/resources/OH-INF/i18n/qbus_nl.properties:
```diff
@@ -0,0 +1,99 @@
+
+
+
+# thing type config description
+thing-type.config.qbus.sample.config1.label = <Your localized config parameter label>
+thing-type.config.qbus.sample.config1.description = <Your localized config parameter description>
+
+# channel types
+channel-type.qbus.sample-channel.label = <Your localized Channel label>
+channel-type.qbus.sample-channel.description = <Your localized Channel description>
```
Please remove these.

fwolter commented on bundles/org.openhab.binding.qbus/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,248 @@
+		<category>Scene</category>
+	</channel-type>
+
+	<channel-type id="co2">
+		<item-type>Number</item-type>
+		<label>CO2</label>
+		<description>CO2 Value for Qbus</description>
+		<category>CO2</category>
+	</channel-type>
+
+	<channel-type id="switch">
+		<item-type>Switch</item-type>
+		<label>Switch</label>
+		<description>Switch Control for Output in Qbus</description>
+		<category>Switch</category>
+	</channel-type>
+
+	<channel-type id="brightness">
+		<item-type>Dimmer</item-type>
+		<label>Brightness</label>
+		<description>Brightness Control for Dimmer Function in Qbus</description>
+		<category>DimmableLight</category>
+	</channel-type>
```
Can these be removed when using the predefined system channels?

fwolter reviewed (changes requested):
There are some checkstyle warnings left. You could take a look at `target/code-analysis/report.html`.

QbusKoen commented:
Hi @fwolter,
Again, i updated the code as requested. Thanks for your patience, i really appreceate this!
I don't know why, but for some reason i can't get the DCO checked.
Can you check my code please, when you find the time for it?
Manu thanks!

fwolter commented on bundles/org.openhab.binding.qbus/src/main/java/org/openhab/binding/qbus/internal/QbusBridgeHandler.java:
```diff
@@ -0,0 +1,313 @@
+                try {
+                    qbusCommunication.startCommunication();
+                } catch (InterruptedException | IOException e) {
+                    logger.warn("Error on restaring communication: {}", e.getMessage());
+                    bridgeOffline("Communication could not be established");
```
The log message could be removed, as the status update is already logged by the framework.

The state change originated by updateStatus() is logged to events.log. Including the status detail message.

Also, it's good practice to include the exception's message in the status detail message to display it in the UI.

fwolter commented on bundles/org.openhab.binding.qbus/src/main/java/org/openhab/binding/qbus/internal/handler/QbusBistabielHandler.java:
```diff
@@ -0,0 +1,210 @@
+     */
+    private void handleSwitchCommand(QbusBistabiel qBistabiel, ChannelUID channelUID, Command command) {
+        if (command instanceof OnOffType) {
+            OnOffType s = (OnOffType) command;
+            String snr = getSN();
+            if (snr != null) {
+                if (s == OnOffType.OFF) {
```
The command don't need to be cast.
```suggestion
            String snr = getSN();
            if (snr != null) {
                if (command == OnOffType.OFF) {
```

fwolter commented on bundles/org.openhab.binding.qbus/src/main/java/org/openhab/binding/qbus/internal/handler/QbusRolHandler.java:
```diff
@@ -0,0 +1,344 @@
+     */
+    private void handleSlatsposCommand(QbusRol qRol, Command command) {
+        String snr = getSN();
+        if (command instanceof org.openhab.core.library.types.UpDownType) {
```
You could add the package to the imports.
```suggestion
        if (command instanceof UpDownType) {
```

fwolter commented on bundles/org.openhab.binding.qbus/src/main/java/org/openhab/binding/qbus/internal/handler/QbusThermostatHandler.java:
```diff
@@ -0,0 +1,244 @@
+        if (command instanceof QuantityType<?>) {
+            QuantityType<?> s = (QuantityType<?>) command;
+            double sp = s.doubleValue();
+            s.toUnit(CELSIUS);
```
You need to use the return value of `toUnit()` for further processing. Keep in mind that it can return null if the value can't be converted.

fwolter commented on bundles/org.openhab.binding.qbus/src/main/java/org/openhab/binding/qbus/internal/protocol/QbusCommunication.java:
```diff
@@ -0,0 +1,925 @@
+    private final Map<Integer, QbusThermostat> thermostat = new HashMap<>();
+    private final Map<Integer, QbusCO2> co2 = new HashMap<>();
+
+    private final ExecutorService threadExecutor = Executors
+            .newSingleThreadExecutor(new NamedThreadFactory("org.openhab.binding.qbus", true));
```
The ID should be the ThingUID. You could add it as an argument to the constructor of `QbusCommunication`.
```suggestion
    private final ExecutorService threadExecutor = Executors
            .newSingleThreadExecutor(new NamedThreadFactory(getThing().getUID().getAsString(), true));
```

fwolter commented on bundles/org.openhab.binding.qbus/src/main/java/org/openhab/binding/qbus/internal/protocol/QbusCommunication.java:
```diff
@@ -0,0 +1,925 @@
+
+        try {
+            startCommunication();
+        } catch (InterruptedException | IOException e) {
+        }
```
The `InterruptedException` shouldn't be cought here, but as high as possible in the program flow to exit the program as directly as possible. You can add `throws InterruptedException` to the method signature to do so.

Nevertheless, the `IOException` should be logged here.

fwolter commented on bundles/org.openhab.binding.qbus/src/main/java/org/openhab/binding/qbus/internal/protocol/QbusCommunication.java:
```diff
@@ -0,0 +1,925 @@
+        BufferedReader reader = qIn;
+
+        if (reader == null) {
+            throw new IOException("Bufferreader for incomming messages not initialized.");
```
```suggestion
            throw new IOException("Bufferreader for incoming messages not initialized.");
```

fwolter commented on bundles/org.openhab.binding.qbus/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,231 @@
+		<description>This bridge represents a Qbus client</description>
+		<config-description>
+			<parameter name="addr" type="text" required="true">
+				<label>IP Address or Host Name</label>
```
Labels are expected to be as short as possible. Guideline is 2-3 words with max 20-25 chars. See https://www.openhab.org/docs/developer/bindings/thing-xml.html#formatting-labels-and-descriptions
```suggestion
				<label>Hostname</label>
```

fwolter commented on bundles/org.openhab.binding.qbus/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,231 @@
+		<config-description>
+			<parameter name="addr" type="text" required="true">
+				<label>IP Address or Host Name</label>
+				<description>IP address of Qbus server, usually 'localhost'</description>
```
```suggestion
				<description>IP address or host name of Qbus server, usually 'localhost'</description>
```

fwolter commented on bundles/org.openhab.binding.qbus/src/main/java/org/openhab/binding/qbus/internal/QbusBridgeHandler.java:
```diff
@@ -0,0 +1,313 @@
+     * @return the ip address
+     */
+    public @Nullable String getAddress() {
+        if (this.config != null) {
+            return this.config.addr;
```
The field can be null at this point, because it isn't inittialized in the constructor. So, you need to check if it is not null before accessing it: if (config != null). But this would result in a compiler warning, as it is a field and could theoretically be modified by another thread exactly between the null check and the access in the next line. To mitigate this, you can store the field to a local variable before checking it:
```
Future<?> localPollingJob = pollingJob;
if (localPollingJob != null) {
    localPollingJob.cancel(true);
}
```

There are some more compiler warnings, which can be seen when calling `mvn clean install`.

fwolter reviewed (changes requested):
The single missing DCO is ok.

QbusKoen commented:
Hi @fwolter 

As usually I updated the changes you requested an pushed them.

2 warnings are left, but i'm not sure if i can do something about them:
[WARNING] C:\Users\ks\Documents\GitHub\openhab-addons\bundles\org.openhab.binding.qbus\pom.xml [0:0]: Unused Export-Package instructions: [org.openhab.*]
[WARNING] C:\Users\ks\Documents\GitHub\openhab-addons\bundles\org.openhab.binding.qbus\pom.xml [0:0]: Unused Import-Package instructions: [io.swagger.v3.oas.annotations.*, javax.annotation.security.*, org.eclipse.jdt.annotation.*, org.openhab.core.automation.annotation.*, com.google.common.*]

Please advice...
Thx!
Kind regards,
Koen

fwolter commented on bundles/org.openhab.binding.qbus/src/main/java/org/openhab/binding/qbus/internal/QbusBridgeHandler.java:
```diff
@@ -0,0 +1,360 @@
+                try {
+                    qbusCommunication.startCommunication();
+                } catch (InterruptedException | IOException e) {
+                    bridgeOffline("Communication could not be established {}" + e.getMessage());
```
The curly brackets don't work with ordinary string concatenation.
```suggestion
                    bridgeOffline("Communication could not be established: " + e.getMessage());
```

fwolter commented on bundles/org.openhab.binding.qbus/src/main/java/org/openhab/binding/qbus/internal/QbusBridgeHandler.java:
```diff
@@ -0,0 +1,360 @@
+                    try {
+                        comm.restartCommunication();
+                    } catch (InterruptedException e) {
+                        bridgeOffline("No connection with Qbus Server" + e.toString());
```
There needs to be a space. Please check all.
```suggestion
                        bridgeOffline("No connection with Qbus Server: " + e.toString());
```

fwolter commented on bundles/org.openhab.binding.qbus/src/main/java/org/openhab/binding/qbus/internal/QbusBridgeHandler.java:
```diff
@@ -0,0 +1,360 @@
+                comm.stopCommunication();
+            } catch (IOException e) {
+                String message = e.toString();
+                logger.error("Error on stopping communication.{} ", message);
```
Bindings should only log to error if something severe happened, like the detection of a bug in your code. This could be debug. See this link for a description of the log levels: https://www.openhab.org/docs/developer/guidelines.html#f-logging

fwolter commented on bundles/org.openhab.binding.qbus/src/main/java/org/openhab/binding/qbus/internal/handler/QbusGlobalHandler.java:
```diff
@@ -0,0 +1,105 @@
+     * @param globalId
+     */
+    public void restartCommunication(QbusCommunication qComm, String type, @Nullable Integer globalId) {
+        try {
+            qComm.restartCommunication();
+        } catch (InterruptedException e) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, e.toString());
+        } catch (IOException e) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, e.toString());
+        }
+
+        if (!qComm.communicationActive()) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, "Communication socket error");
+            return;
+        }
+
+        QbusBridgeHandler qBridgeHandler = getBridgeHandler(type, globalId);
+        if (qBridgeHandler != null) {
+            qBridgeHandler.bridgeOnline();
+        }
```
You might want to call `qBridgeHandler.bridgeOnline()` only if no exception occurred. Otherwise, the Thing will be set to OFFLINE and then, `qBridgeHandler.bridgeOnline()` will be called.

fwolter commented on bundles/org.openhab.binding.qbus/src/main/java/org/openhab/binding/qbus/internal/handler/QbusSceneHandler.java:
```diff
@@ -0,0 +1,229 @@
+    void handleSwitchCommand(QbusScene qScene, ChannelUID channelUID, Command command) {
+        String snr = getSN();
+        if (command instanceof OnOffType) {
+            OnOffType s = (OnOffType) command;
+            if (s == OnOffType.OFF) {
```
The command don't need to be cast. Please check all.

fwolter commented on bundles/org.openhab.binding.qbus/src/main/java/org/openhab/binding/qbus/internal/handler/QbusThermostatHandler.java:
```diff
@@ -0,0 +1,265 @@
+                if (spCelcius != null) {
+                    qThermostat.executeSetpoint(sp, snr);
+                } else {
+                    thingOffline("Could not set setpoint for thermostat (convertion failed)  " + thermostatId);
```
As this is not a permanent error, you are good to log it only.
```suggestion
                    thingOffline("Could not set setpoint for thermostat (conversion failed)  " + thermostatId);
```

fwolter commented on bundles/org.openhab.binding.qbus/src/main/java/org/openhab/binding/qbus/internal/protocol/QbusCommunication.java:
```diff
@@ -0,0 +1,949 @@
+            startCommunication();
+        } catch (InterruptedException | IOException e) {
+            String message = e.toString();
+            logger.error("Could not start the communication with the server. {}", message);
```
Bindings should only log to error if something severe happened, like the detection of a bug in your code. This could be debug. See this link for a description of the log levels: https://www.openhab.org/docs/developer/guidelines.html#f-logging

fwolter commented on bundles/org.openhab.binding.qbus/src/main/java/org/openhab/binding/qbus/internal/protocol/QbusCommunication.java:
```diff
@@ -0,0 +1,949 @@
+            logger.warn("Error sending message, trying to restart communication");
+            try {
+                restartCommunication();
+            } catch (InterruptedException e) {
+                if (handler != null) {
+                    handler.bridgeOffline("No communication with Qbus server. " + e.toString());
+                }
```
As `sendMessage()` already throws the `InterruptedException`, you could simply remove this catch clause.

fwolter commented on bundles/org.openhab.binding.qbus/src/main/java/org/openhab/binding/qbus/internal/protocol/QbusCommunication.java:
```diff
@@ -0,0 +1,949 @@
+                    if (handler != null) {
+                        handler.bridgeOffline("Could not request outputs from client. {}" + message);
+                    } else {
+                        logger.error("Could not request outputs from client. {}", message);
```
See above. Logging to error. Please check all.

fwolter commented on bundles/org.openhab.binding.qbus/src/main/java/org/openhab/binding/qbus/internal/protocol/QbusCommunication.java:
```diff
@@ -0,0 +1,949 @@
+
+                return;
+            }
+        } else
+
+        {
```
Please remove this empty line.

QbusKoen commented:
@fwolter 
Check! Updated requested changes.

QbusKoen commented:
Hi @fwolter, any news on the update yet? Should I perform further actions for this Binding?
Thankx in advance for answering.

fwolter commented on bundles/org.openhab.binding.qbus/src/main/java/org/openhab/binding/qbus/internal/handler/QbusRolHandler.java:
```diff
@@ -0,0 +1,358 @@
+            localPollingJob.cancel(true);
+        }
+
+        if (localPollingJob == null || localPollingJob.isCancelled()) {
+            this.config = getConfig().as(QbusThingsConfig.class);
+        }
+        this.config = getConfig().as(QbusThingsConfig.class);
```
Does this make any sense?

fwolter commented on bundles/org.openhab.binding.qbus/src/main/java/org/openhab/binding/qbus/internal/handler/QbusSceneHandler.java:
```diff
@@ -0,0 +1,228 @@
+            localPollingJob.cancel(true);
+        }
+
+        if (localPollingJob == null || localPollingJob.isCancelled()) {
+            this.config = getConfig().as(QbusThingsConfig.class);
+        }
+        this.config = getConfig().as(QbusThingsConfig.class);
```
Same here.

fwolter commented on bundles/org.openhab.binding.qbus/src/main/java/org/openhab/binding/qbus/internal/protocol/QbusCommunication.java:
```diff
@@ -0,0 +1,943 @@
+        if (bridgeCallBack != null) {
+            if (ctdConnected) {
+                try {
+                    sendAndReadMessage("getBistabiel");
```
`initialize()` must return fast. As this will call some network code, you need to execute it asynchronously e.g. via `scheduler.submit()`.

QbusKoen commented:
Dear @fwolter,
I took some time to rework this binding. You will probably notice a lot of changes. When i test the binding now, it works great ;-). 
Can you be so kind to do antother review please?

fwolter commented on bundles/org.openhab.binding.qbus/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,232 @@
+				<default>8447</default>
+				<advanced>true</advanced>
+			</parameter>
+			<parameter name="serverCheck" type="integer" required="false">
```
```suggestion
			<parameter name="serverCheck" type="integer" required="false" unit="min" min="1">
```

fwolter commented on bundles/org.openhab.binding.qbus/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,232 @@
+			</parameter>
+			<parameter name="serverCheck" type="integer" required="false">
+				<label>Server Check</label>
+				<description>Time to check communication with Qbus Server (min), default 10. If set to 0 or left empty, no
+					refresh
+					will be scheduled</description>
```
Can you fix the line break? Put all into one line and then run `mvn spotless:apply`.

QbusKoen commented on bundles/org.openhab.binding.qbus/src/main/java/org/openhab/binding/qbus/internal/QbusBridgeHandler.java:
```diff
@@ -0,0 +1,380 @@
+        }
+
+        try {
+            addr = InetAddress.getByName(getAddress());
```
Thanks for pointing this out. I removed port and addr as they are set in the communications object.

QbusKoen commented:
@fwolter as requested i updated the binding.
Can you do another check when you find the time?
Many thankx!

fwolter commented on bundles/org.openhab.binding.qbus/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -24,11 +24,11 @@
 				<description>Time to check communication with Qbus Server (min), default 10. If set to 0 or left empty, no
 					refresh
-					will be scheduled</description>
+					willl be scheduled</description>
```
The line breaks still seem corrupted.

QbusKoen replied:
Hmm, strange. When i fix the line break and do mvn spotless:apply the line break comes back.
When I do a build without the linebreak, i get the error that i should do mvn spotless:apply.

fwolter replied:
The trick is to put the entire tag in one line and then run spotless:apply.

QbusKoen replied:
Yes, that's what i did, did it again:
<description>Time to check communication with Qbus Server (min), default 10. If set to 0 or left empty, no refresh will be scheduled</description>
_mvn spotless:apply -pl :org.openhab.binding.qbus_
<description>Time to check communication with Qbus Server (min), default 10. If set to 0 or left empty, no refresh
					will be scheduled</description>
**Then i putted the whole parameter in one line:**
<parameter name="serverCheck" type="integer" required="false" unit="min" min="1"><label>Server Check</label><description>Time to check communication with Qbus Server (min), default 10. If set to 0 or left empty, no refresh will be scheduled</description><default>10</default><advanced>true</advanced></parameter>
**And after the spotless, same result:**
			<parameter name="serverCheck" type="integer" required="false" unit="min" min="1">
				<label>Server Check</label>
				<description>Time to check communication with Qbus Server (min), default 10. If set to 0 or left empty, no refresh
					will be scheduled</description>
				<default>10</default>
				<advanced>true</advanced>
			</parameter>

I even copied the line to notepad++ where you can vusualize line breaks. I'm sure they where not there. Copied again to Eclipse. Same relut.!?

fwolter replied:
This result with a single line break is just right:
```
<description>Time to check communication with Qbus Server (min), default 10. If set to 0 or left empty, no refresh
					will be scheduled</description>
```

QbusKoen replied:
oh, ok. So it's time to commit again?

fwolter replied:
Yes!

QbusKoen replied:
:-)

fwolter reviewed (approved):
LGTM

Now it's time to wait for another review, before this can be merged. Please don't add any new features in the meantime.

QbusKoen commented:
Great, thank you very much for your patience and your great advicei I really appreciete this!
What do you mean with another review? Does one of your collegue's will check this to? Or do I have to ask users to test the binding and give comments?

fwolter commented:
Every new addon needs to be reviewed by two openHAB addons maintainers. The `cre` label indicates that a second review is pending. You probably need some patience as the backlog is quite full.

fwolter commented:
Due to lack of review capacity, the rules for new bindings were weakened. New bindings need only the approval of one reviewer instead of two from now on.

Confectrician commented:
I think this PR broke the website build, since it attempts to load a huge logo file in a doc folder.

Maybe the codeowner can remove this jpg from the readme (and repo)  and provide a **smaller** (not 3500 x 2400 px please) logo picture as a pull request on the openHAB docs repository.

Here is an example of how to add a logo to the documentation properly:
https://github.com/openhab/openhab-docs/pull/1523

The file must be named like the bundle name. "qbus" in this case.
